---
output:
  md_document:
    variant: markdown_github
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, echo = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-"
)
```

# forrel <img src="man/figures/logo.png" align="right" height=140/>

## Introduction
The goal of `forrel` is to provide forensic pedigree computations and relatedness inference from genetic marker data. Some of its functionality is derived from `paramlink`, which is no longer actively developed. 'forrel' is part of the _ped suite_, a collection of packages for pedigree analysis with `pedtools` as the central package for handling pedigrees and marker data.

The most important analyses currently supported by forrel are:

* Pairwise relatedness inference: Estimation of IBD (kappa) coefficients from marker data
* Visualisation of IBD coefficients in the IBD triangle
* Likelihood ratios for relationship testing
* Simulation of marker genotypes and complete DNA profiles. Unconditional or conditional on known genotypes
* Power analysis for relationship testing: Exclusion power (EP) and inclusion power (IP)
* Tailor-made functions for power analysis in family reunion cases: 
    * `missingPersonPlot()`
    * `missingPersonEP()`
    * `missingPersonIP()`
    * `MPPsims()`
    * `powerPlot()`
* Import of pedigree data and frequency databases from the `Familias` software.

## Installation
To get the latest version, install from GitHub as follows:

```{r, eval = FALSE}
 # First install devtools if needed
if(!require(devtools)) install.packages("devtools")

# Install forrel from GitHub
devtools::install_github("magnusdv/forrel")
```

## An example
In this short introduction, we first demonstrate simulation of marker data for a pair of siblings. Then - pretending the relationship is unknown to us - we estimate the relatedness between the brothers using the simulated data. If all goes well, the estimate should be close to the expected value for siblings.

To get going, load the `forrel` and `pedtools` packages. The latter is needed for pedigree creation and plotting. 
```{r}
library(pedtools)
library(forrel)
```

**Create the pedigree**  

We start by creating and plotting a pedigree with two brothers, named `bro1` and `bro2`. 

```{r, eval=FALSE}
bros = c("bro1", "bro2")
x = nuclearPed(children = bros)
plot(x)
```
```{r sibs, echo=FALSE, fig.height=2.5, fig.width=2.5, fig.align="center"}
bros = c("bro1", "bro2")
x = nuclearPed(children = bros)
plot(x, margins = c(1,1,1,1))
```


**Marker simulation**  

Now let us simulate the genotypes of 100 independent SNPs for the brothers. Each SNP has alleles 1 and 2, with equal frequencies by default. This is an example of _unconditional_ simulation, since we don't give any genotypes to condition on. Unconditional simulation is performed by simple _gene dropping_, i.e., by drawing random alleles independently for the parents, followed by a "Mendelian coin toss" in each parent-child transmission.
```{r}
x = markerSim(x, N = 100, ids = bros, alleles = 1:2, seed = 1234)
```
Note 1: The `seed` argument is passed onto the random number generator. If you use the same seed, you should get exactly the same results.  
Note 2: To suppress the informative messages printed during simulation, add `verbose = FALSE` to the function call.

The pedigree `x` now has 100 markers attached to it. The genotypes of the first few markers are shown when printing `x` to the screen:
```{r}
x
```

**Estimation of IBD coefficients**  

The `IBDestimate()` function estimates the coefficients of _identity-by-descent_ (IBD) between pairs of individuals, from the available marker data. 
```{r}
k = IBDestimate(x, ids = bros)
k
```

The theoretical expectation for non-inbred full siblings is $(\kappa_0, \kappa_1, \kappa_2) = (0.25, 0.5, 0.25)$.
To get a visual sense of how close our estimate is, it is instructive to plot it in the IBD triangle:
```{r triangle, fig.height=4, fig.width=4, fig.align="center"}
showInTriangle(k, labels = TRUE)
```
